### ISSUE TITLE
2.6.2 - Potential Paginator Bug

### ISSUE THREAD

**Issue by snsmurf** [cakephp/cakephp#5973]
Slight issue with the Paginator component. If you override the default limit and set 25 for example, and then proceed to amend the limit via a named parameter, the maxLimit will actually be set to the new default and not actually set to 100 like it defaults to in the Paginator component settings so now no matter how many you wish to retrieve it always returns 25.

So more step form to reproduce, we have 

	$this->paginate = array(
		'limit' => 25
	);
	$data = $this->paginate();

We then have a drop down to allow the client to amend the limit on the fly to various increments (10, 25, 50, 100). Which gives us a url like:

    /blog_posts/index/limit:50

Paginate will still just return 25.
Debugging the paginator options at various points throughout the process makes it seems like it could be [this line](https://github.com/cakephp/cakephp/blob/master/lib/Cake/Controller/Component/PaginatorComponent.php#L339). But why it's doing that, it just seems that if you override the default limit you're always going to run into this issue.

**Comment by dereuromark** [comment]
You are using the wrong key, see the documentation about `maxLimit`: http://book.cakephp.org/2.0/en/core-libraries/components/pagination.html#limit-the-maximum-number-of-rows-that-can-be-fetched

Side-note: One shouldn't use named params anymore since 2.0.0 :)
Query strings are the correct approach.

**Comment by snsmurf** [comment]
I don't see how the maxLimit section there applies?

I've amended the default 'limit' value, and in turn that ends up changing the default 'maxLimit' from 100 to 25 (which is my default limit), which then prevents any changing of the limit to be over 25 via named or query string parameters.

**Comment by ADmad** [comment]
@snsmurf  [The line](https://github.com/cakephp/cakephp/blob/master/lib/Cake/Controller/Component/PaginatorComponent.php#L339)  you pointed out would be executed only if maxLimit is not set or limit > maxLimit. So you must be having some other code which either unsets or sets maxLimit to a lower value.

**Comment by snsmurf** [comment]
So for demo purposes
	public function getDefaults($alias) {
		$defaults = $this->settings;
		if (isset($this->settings[$alias])) {
			$defaults = $this->settings[$alias];
		}
		debug($defaults);
		if (isset($defaults['limit']) &&
			(empty($defaults['maxLimit']) || $defaults['limit'] > $defaults['maxLimit'])
		) {
			$defaults['maxLimit'] = $defaults['limit'];
		}
		return array_merge(
			array('page' => 1, 'limit' => 20, 'maxLimit' => 100, 'paramType' => 'named'),
			$defaults
		);
	}
That's the get defaults method now with my debug in. That debug returns this:

	array(
		'limit' => (int) 25,
		'conditions' => array(),
		'contain' => array(
			(int) 0 => 'BlogCategory'
		)
	)	

So from that perspective, it's going to hit that line and amended the maxLimit and set it to 25 causing my issue.

All I have done is override the default limit to 25. I have done a full search over my whole project for maxLimit as well just to double check and I can confirm that I am not doing anything with maxLimit so everything that is happening is out of box with the paginator component

**Comment by ADmad** [comment]
Okay, so the problem seems to occur when `maxLimit` is not set in `$this->settings[$alias]`.  Did you set paginate as 
```
$this->paginate = array('BlogPost' => array('limit' => 25));
```` 
or 
```
$this->paginate = array('limit' => 25);
```
?

**Comment by ADmad** [comment]
Nevermind, the problem can occur in any case when `maxLimit` is not specified in setting. So Paginator should either throw an error is maxLimit is not set or just set 100 as default.

**Comment by snsmurf** [comment]
Originally the second way, I have just tested the first way however and the same occurs for both ways. It seems to be just a simple case of paginator not merging it's defaults with what I have set.

**Comment by markstory** [comment]
I can take a look at this. I don't think we should be setting maxLimit is it is empty on line 337.

**Comment by markstory** [comment]
Fixed in #5981 I agree that having `maxLimit` infer its value based on limit being defined was pretty opaque.
